src/wagtail_vector_index/consumers.py

```diff
@@ -0,0 +1,149 @@
+        process_prompt: Processes the incoming prompt and sends SSE updates.
+        check_permissions: Checks user authentication and raises PermissionDenied if not authenticated.
+        ratelimit_request: Placeholder for implementing rate-limiting logic.
+
+    Usage:
+    - Inherit from this class and set either 'page_instance' or 'page_model_name' attribute.
+    - Implement custom logic within 'process_prompt' to handle vector index queries and SSE updates.
+    - Optionally, override 'check_permissions' and 'ratelimit_request' for additional safety checks.
+
+    Example:
+    ```python
+    class CustomSSEConsumer(WagtailVectorIndexSSEConsumer):
+        page_instance = YourWagtailPageModel
+
+        async def process_prompt(self, query):
+            # Your custom logic to handle the query and send SSE updates
+            pass
+    ```
+    """
+    page_instance = None
+    page_model_name = None
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initializes the consumer and checks the required attributes.
+
+        Raises:
+            ValueError: If neither 'page_instance' nor 'page_model_name' is set.
+            ValueError: If the specified 'page_model_name' is not found.
+            ValueError: If the specified page model does not inherit the ModelVectorIndex mixin.
+        """
+        super().__init__(*args, **kwargs)
+
+        # Check if either page_model_name or page_instance is set
+        if self.page_model_name is None and self.page_instance is None:
+            raise ValueError('You must set either the page_model_name or page_instance attribute')
+
+        if not self.page_instance:
+            try:
+                self.page_instance = apps.get_model(self.page_model_name)
+            except LookupError:
+                raise ValueError(f'Model {self.page_model_name} not found')
+
+        # Check if the page model has the required method (ModelVectorIndex mixin)
+        if not hasattr(self.page_instance, 'get_vector_index') or not callable(self.page_instance.get_vector_index):
+            raise ValueError('Your page_model must inherit the ModelVectorIndex mixin')
+
+        self.vector_index = self.page_instance.get_vector_index()
+
+    async def handle(self, body):
+        """
+        Handles HTTP requests, sets up SSE headers, and processes prompts.
+
+        Raises:
+            PermissionDenied: If the user is not authenticated.
+        """
+        # Send SSE headers
+        await self.send_headers(headers=[
+            (b"Cache-Control", b"no-cache"),
+            (b"Content-Type", b"text/event-stream"),
+            (b"Transfer-Encoding", b"chunked"),
+        ])
+
+        try:
+            try:
+                user = self.scope['user']
+            except KeyError as e:
+                raise ValueError('User not found in scope, make sure AuthMiddlewareStack is applied correctly') from e
+            
+            await self.check_permissions(user)  # Check permissions
```

Comment:
**suggestion:** These (`check_permissions` and `ratelimit_request`) don't need to be the responsibility of this consumer and could instead be implemented by the user as a higher-order function/middleware, e.g. `AuthMiddlewareStack` from `channels`.

Reply:
I don't think we need to implement anything to do this, at least for v1